safe_control_gym/controllers/mpc/mpc_acados.py

```diff
@@ -0,0 +1,412 @@
+            env_func,
+            horizon=horizon,
+            q_mpc=q_mpc,
+            r_mpc=r_mpc,
+            warmstart=warmstart,
+            soft_constraints=soft_constraints,
+            soft_penalty=soft_penalty,
+            terminate_run_on_done=terminate_run_on_done,
+            constraint_tol=constraint_tol,
+            output_dir=output_dir,
+            additional_constraints=additional_constraints,
+            use_gpu=use_gpu,
+            seed=seed,
+            **kwargs
+        )
+
+        self.x_guess = None
+        self.u_guess = None
+        # acados settings
+        self.use_RTI = use_RTI
+
+    @timing
+    def reset(self):
+        print(colored('Resetting MPC', 'green'))
+        '''Prepares for training or evaluation.'''
+        # Setup reference input.
+        if self.env.TASK == Task.STABILIZATION:
+            self.mode = 'stabilization'
+            self.x_goal = self.env.X_GOAL
+        elif self.env.TASK == Task.TRAJ_TRACKING:
+            self.mode = 'tracking'
+            self.traj = self.env.X_GOAL.T
+            # Step along the reference.
+            self.traj_step = 0
+
+        self.acados_model = None
+        self.ocp = None
+        self.acados_ocp_solver = None
+        # Dynamics model.
+        self.set_dynamics_func()
+        self.setup_acados_model()
+        # Acados optimizer.
+        self.setup_acados_optimizer()
+        self.acados_ocp_solver = AcadosOcpSolver(self.ocp)
+        # Previously solved states & inputs, useful for warm start.
+        self.x_prev = None
+        self.u_prev = None
+        self.setup_results_dict()
+
+    def setup_acados_model(self) -> AcadosModel:
+
+        acados_model = AcadosModel()
+        acados_model.x = self.model.x_sym
+        acados_model.u = self.model.u_sym
+        acados_model.name = self.env.NAME
+
+        # set up rk4 (acados need symbolic expression of dynamics, not function)
+        k1 = self.model.fc_func(acados_model.x, acados_model.u)
+        k2 = self.model.fc_func(acados_model.x + self.dt / 2 * k1, acados_model.u)
+        k3 = self.model.fc_func(acados_model.x + self.dt / 2 * k2, acados_model.u)
+        k4 = self.model.fc_func(acados_model.x + self.dt * k3, acados_model.u)
+        f_disc = acados_model.x + self.dt / 6 * (k1 + 2 * k2 + 2 * k3 + k4)
+
+        acados_model.disc_dyn_expr = f_disc
+        '''
+        Other options to set up the model:
+        f_expl = self.model.x_dot (explicit continuous-time dynamics)
+        f_impl = self.model.x_dot_acados - f_expl (implicit continuous-time dynamics)
+        model.f_impl_expr = f_impl
+        model.f_expl_expr = f_expl
+        '''
+        # store meta information # NOTE: unit is missing
+        acados_model.x_labels = self.env.STATE_LABELS
+        acados_model.u_labels = self.env.ACTION_LABELS
+        acados_model.t_label = 'time'
+
+        self.acados_model = acados_model
+
+    @timing
+    def compute_initial_guess(self, init_state, goal_states):
```

Comment:
Is this the same as the function in MPC? why is it being overridden?

Reply:
This is updated with using super to reduce some duplicated parts